src/Microsoft.Identity.Client/Telemetry.cs

```diff
@@ -48,5 +51,137 @@ public static Telemetry GetInstance()
+            {
+                // if this is the first event associated to this
+                // RequestId we need to initialize a new List to hold
+                // all of sibling events
+                List<EventBase> events = new List<EventBase>();
+                events.Add(eventToStop);
+                CompletedEvents[requestId] = events;
+            }
+            else
+            {
+                // if this event shares a RequestId with other events
+                // just add it to the List
+                CompletedEvents[requestId].Add(eventToStop);
+            }
+
+            // Mark this event as no longer in progress
+            EventsInProgress.Remove(eventKey);
+        }
+
+        internal void Flush(string requestId)
+        {
+            if (null == _receiver) return;
+
+            // check for orphaned events...
+            List<EventBase> orphanedEvents = CollateOrphanedEvents(requestId);
+            // Add the OrphanedEvents to the completed EventList
+            if (!CompletedEvents.ContainsKey(requestId))
+            {
+                // No completed Events returned for RequestId
+                return;
+            }
+
+            CompletedEvents[requestId].AddRange(orphanedEvents);
+
+            List<EventBase> eventsToFlush = CompletedEvents[requestId];
+            CompletedEvents.Remove(requestId);
+
+            if (TelemetryOnFailureOnly)
+            {
+                // iterate over Events, if the ApiEvent was successful, don't dispatch
+                bool shouldRemoveEvents = false;
+
+                foreach (var e in eventsToFlush)
+                {
+                    var apiEvent = e as ApiEvent;
+                    if (apiEvent != null)
+                        shouldRemoveEvents = apiEvent.WasSuccessful;
+                    break;
+                }
+
+                if (shouldRemoveEvents)
+                {
+                    eventsToFlush.Clear();
+                }
+            }
+
+            if (eventsToFlush.Count > 0)
+            {
+                var dictionariesToFlush = eventsToFlush.Cast<Dictionary<string, string>>().ToList();
+                dictionariesToFlush.Insert(0, BuildDefaultEvent());
+                _receiver(dictionariesToFlush);
+            }
+        }
+
+        private List<EventBase> CollateOrphanedEvents(String requestId)
+        {
+            var orphanedEvents = new List<EventBase>();
+            foreach (var key in EventsInProgress.Keys)
+            {
+                if (key.Item1 == requestId)
+                {
+                    // The orphaned event already contains its own start time, we simply collect it
+                    orphanedEvents.Add(EventsInProgress[key]);  
+                    EventsInProgress.Remove(key);  // remove() this entry (clean up!)
+                }
+            }
+            return orphanedEvents;
+        }
+
+        private Dictionary<string, string> BuildDefaultEvent()
```

Comment:
i would still recommend moving default event to its own class. Telemetry class should not become a container for events. Having the event name blank or event adding "Default Event" is not that bad, is it?

---
In reply to: [111027529](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/314#discussion_r111027529) [](ancestors = 111027529)